rskj-core/src/main/java/org/ethereum/core/BlockHeader.java

```diff
@@ -539,28 +533,24 @@ public String getParentShortHash() {
     public byte[] getMiningForkDetectionData() {
         if(includeForkDetectionData) {
             if (hasMiningFields() && miningForkDetectionData.length == 0) {
-                byte[] encodedBlock = getEncoded(false, false);
-                byte[] hashForMergedMining = HashUtil.keccak256(encodedBlock);
-
-                byte[] hashForMergedMiningPrefix = Arrays.copyOfRange(
-                        hashForMergedMining,
-                        0,
-                        HASH_FOR_MERGED_MINING_PREFIX_LENGTH
-                );
+                byte[] hashForMergedMining = getBaseHashForMergedMining();
+
                 byte[] coinbaseTransaction = getBitcoinMergedMiningCoinbaseTransaction();
 
-                List<Byte> hashForMergedMiningPrefixAsList = Arrays.asList(ArrayUtils.toObject(hashForMergedMiningPrefix));
-                List<Byte> coinbaseAsList = Arrays.asList(ArrayUtils.toObject(coinbaseTransaction));
+                byte[] mergeMiningTagPrefix = Arrays.copyOf(RskMiningConstants.RSK_TAG, RskMiningConstants.RSK_TAG.length + HASH_FOR_MERGED_MINING_PREFIX_LENGTH);
+                arraycopy(hashForMergedMining, 0, mergeMiningTagPrefix, RskMiningConstants.RSK_TAG.length, HASH_FOR_MERGED_MINING_PREFIX_LENGTH);
 
-                int position = Collections.lastIndexOfSubList(coinbaseAsList, hashForMergedMiningPrefixAsList);
+                int position = ListArrayUtil.lastIndexOfSubList(coinbaseTransaction, mergeMiningTagPrefix);
                 if (position == -1) {
                     throw new IllegalStateException(
                             String.format("Mining fork detection data could not be found. Header: %s", getShortHash())
                     );
                 }
 
-                int from = position + HASH_FOR_MERGED_MINING_PREFIX_LENGTH;
+                int from = position + RskMiningConstants.RSK_TAG.length + HASH_FOR_MERGED_MINING_PREFIX_LENGTH;
                 int to = from + FORK_DETECTION_DATA_LENGTH;
+
+                // should the fork detection data in the coinbase be under 12 bytes, copyOfRange pads with zeros
```

Comment:
I think we should definitely validate that the fork detection data in the coinbase transaction is always 12 bytes and avoid these potentially consensus threatening blocks.

Reply:
Done